### ISSUE TITLE
Max clock limit including offset

### ISSUE THREAD

**Issue by JohanAR** [xor2k/gpu_undervolt#1]
`nvidia-smi -i $gpu -lgc 100,$((gpu_high - offset))`

I assume this is because Nvidia settings shows graphics clock max including the OC offset. However, as far as I could see while testing, the actual clock limit is whatever you set with lgc. I.e. if I run

```
nvidia-smi -i 0 -lgc 210,1900
nvidia-settings -a "[gpu:0]/GPUGraphicsClockOffsetAllPerformanceLevels=400"
```

Then the actual GPU clock does not exceed 1900 even though Nvidia settings reports a much high max clock.

**Comment by xor2k** [comment]
As far as I understood it, the offset is some sort of "overclocking with undervolting": higher clock speed without increasing voltage. For pure undervolting purposes, one would just end up at the same MaxClock as before.

Maybe I got something wrong. But in the end, it does not really matter. What matters is to have good benchmarks to validate different sorts of ClockOffset/No ClockOffset combinations. Unfortunately, I could not find any working AI benchmark, seems all quite broken right now. Do you know a good one?

**Comment by xor2k** [comment]
Is there anything that can be done here?

**Comment by JohanAR** [comment]
I think offset is purely overclocking, and the undervolting is a side-effect of the power limit we set with nvidia-smi. 

With undervolting (or power limiting as we do on Linux) I get a max clock speed that is about 5-10% lower than what I had with stock settings, but it could of course be adjusted for less performance loss in exchange for higher power usage. I can live with slightly less performance if I save 150W power usage :)

I used regular GPU benchmarks and played games to verify my settings. Did some AI work later but did not experience any instability, so the OC I got should be safe. But to answer your question, I don't know of any AI benchmarks. Maybe just run stable diffusion and generate large images with high float precision?

**Comment by xor2k** [comment]
I did run stable diffusion actually, yeah, that's a good benchmark. I've also observed that it's 5-10% slower while saving around 100W for me. Just thought whether you had some idea on how to improve the current way it's done in the script or had some more insight. Accoring to the `nvidia-smi` docs, this is what `-lgc` does:
```
   -lgc, --lock-gpu-clocks=MIN_GPU_CLOCK,MAX_GPU_CLOCK
       Specifies  <minGpuClock,maxGpuClock>  clocks as a pair (e.g. 1500,1500)
       that defines closest desired locked GPU clock speed in MHz.  Input  can
       also  use  be  a  singular  desired clock value (e.g. <GpuClockValue>).
       Supported on Volta+.  Requires root
```
So either it works that way or it does not 😅

**Comment by JohanAR** [comment]
I put the values I found for my GPU in a script that I run when I want to apply the limit. Not sure how far you can go with a generic script that works for everybody, since exact values will depend on how lucky you are with the card you have, and you have to do some experimenting to find the max clocks at a given voltage.

```
#!/bin/sh

set -e

# Set persistence mode on
nvidia-smi -i 0 -pm 1

# Set 200W power limit
nvidia-smi -i 0 -pl 200

# Limit max clock speed to 1875 MHz
nvidia-smi -i 0 -lgc 210,1850

sleep 1

# Add 350 MHz
nvidia-settings -a "[gpu:0]/GPUGraphicsClockOffsetAllPerformanceLevels=325"
```

Have you used MSI Afterburner on Windows? It allows you to exactly define the voltage-clock table, which I think better shows how the GPU actually works. Power limit + overlock, which we have to do on Linux, has the same end result but it's a more complicated way to go about it.

**Comment by xor2k** [comment]
Yes, it's more complicated on Linux. Any reason to add the `sleep 1`?

The power limit should not be necessary, unless your cooling capacity is limited. Once the clocks are fixed, power draw should not exceed a certain value anyway.

Indeed, some luck is involved. However, margins for stable operation are pretty huge, so if it works for 99% of the people out of the box and each can save 100W and 1% needs to make some modifications (or probably just leave it alone), should be fine, shouldn't it?

I can add your values to the current list. What's your GPU ID? You can run
```
nvidia-smi --query-gpu=gpu_name --format=csv,noheader | sed -e 's/ /_/g'
```
to find out.

**Comment by JohanAR** [comment]
IIRC it was because increasing clock speed by 350 would sometimes happen faster than limiting max clock speed to 1875, which caused some freezes if I had a game running when I launched the script.

I did a quick test removing the power limit from the script and it hovers mostly in the 205-220W range (just tested 1 game). Perhaps it's generally safer to not set a limit? I don't know. But now that I have something that is stable for me, I don't mind saving those few extra watts.

I haven't tried to _really_ push the limits, but as you can see from the comments in my script not matching the actual commands I reduced both max limit and clock offset by 25MHz at the end.. Thought 1875/350 were stable but then I got a freeze when playing a new game. If the defaults are meant to save 100W you ought to be able to have higher values though

GPU ID is `NVIDIA_GeForce_RTX_3080`

**Comment by xor2k** [comment]
Hi there, any news? Otherwise I might close this issue.

**Comment by JohanAR** [comment]
Sorry, nothing new. In the end I just kept using the values I found stable and forgot about it. I'll close it.
